### ISSUE TITLE
DOMParser doesn't work in translators in sandbox

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#1502]
Moving from https://github.com/zotero/scaffold/issues/74, which has an example.

I doubt we'll be able to fix this in the current codebase. As far as I can tell, there's no easy way to access a functional DOMParser in a sandbox in current versions of Firefox, and while there may be a way to do it, pretty much the only search result I can find for it is [someone telling @simonster it won't work](https://groups.google.com/forum/#!topic/mozilla.dev.platform/w3ZX8WdijuU). (And the previous method he was using to do this doesn't work anymore.)

This doesn't affect the connector, and as far as I know it doesn't affect any of the translators we use for Add Item by Identifier or Retrieve Metadata for PDF, though it affects saving from feeds. It also affects translation-server, but we're hoping to rewrite that in node.js soon. Finally, it means we can't use translators more widely in the client (e.g., for metadata updating or PDF retrieval), so those may have to wait for Electron. We can also see if things are any different in Firefox 60.

FWIW, in case anyone wants to wrestle with this, I got a bit farther with this code, replacing the [current DOMParser shim](https://github.com/zotero/zotero/blob/763d7dec75c797bd49d7ddc51e313056d6b736ce/chrome/content/zotero/xpcom/translation/translate_firefox.js#L388):

```javascript
let parser = Components.classes["@mozilla.org/xmlextras/domparser;1"]
	.createInstance(Components.interfaces.nsIDOMParser);
let dp = function () {
	var obj = {
		parseFromString: function (str, contentType) {
			return parser.parseFromString(str, contentType);
		}
	};
	return Components.utils.cloneInto(obj, sandbox, {cloneFunctions: true});
};
sandbox.DOMParser = Components.utils.cloneInto(dp, sandbox, {cloneFunctions: true});
```
But then I just got a permissions error on `querySelector` from the returned document. Setting the principal and URIs with the DOMParser's `init()` method (as in a [previous version](https://github.com/zotero/zotero/blob/3d002674f153322623f4fab8e2d1fd936e414b5d/chrome/content/zotero/xpcom/translation/translate_firefox.js#L63-L69) of the code) didn't help.

(This can be tested easily using translation-server with the Zotero code mounted as an external volume, as explained in the instructions.)

**Comment by dstillman** [comment]
This breaks PDF retrieval for ScienceDirect pages and possibly others.

https://forums.zotero.org/discussion/comment/321342/#Comment_321342

We have a fix in https://github.com/zotero/zotero/commit/c201ce00d967d3b211064be3f9009ea07ab08ace, but we have to make sure we really want to do that.

**Comment by dstillman** [comment]
Fixed by 5216f1188531627fd7543d2d149b72a9301622c5
